src/compile/format.ts

```diff
@@ -1,114 +1,150 @@
+      config.timeFormat,
+      isScaleFieldDef(fieldOrDatumDef) && fieldOrDatumDef.scale?.type === ScaleType.UTC
     );
     return signal ? {signal} : undefined;
-  } else if (!omitNumberFormatAndEmptyTimeFormat) {
-    format = numberFormat(channelDefType(fieldOrDatumDef), format, config);
-    if (isFieldDef(fieldOrDatumDef) && isBinning(fieldOrDatumDef.bin)) {
-      const endField = vgField(fieldOrDatumDef, {expr, binSuffix: 'end'});
-      return {
-        signal: binFormatExpression(field, endField, format, formatType, config)
-      };
-    } else if (format || channelDefType(fieldOrDatumDef) === 'quantitative') {
-      return {
-        signal: `${formatExpr(field, format)}`
-      };
+  }
+
+  format = numberFormat(channelDefType(fieldOrDatumDef), format, config);
+  if (isFieldDef(fieldOrDatumDef) && isBinning(fieldOrDatumDef.bin)) {
+    const endField = vgField(fieldOrDatumDef, {expr, binSuffix: 'end'});
+    return {
+      signal: binFormatExpression(field, endField, format, formatType, config)
+    };
+  } else if (format || channelDefType(fieldOrDatumDef) === 'quantitative') {
+    return {
+      signal: `${formatExpr(field, format)}`
+    };
+  } else {
+    return {signal: `isValid(${field}) ? ${field} : ""+${field}`};
+  }
+}
+
+function fieldToFormat(
+  fieldOrDatumDef: FieldDef<string> | DatumDef<string>,
+  expr: 'datum' | 'parent' | 'datum.datum',
+  normalizeStack: boolean
+) {
+  if (isFieldDef(fieldOrDatumDef)) {
+    if (normalizeStack) {
+      return `${vgField(fieldOrDatumDef, {expr, suffix: 'end'})}-${vgField(fieldOrDatumDef, {
+        expr,
+        suffix: 'start'
+      })}`;
     } else {
-      return {signal: `isValid(${field}) ? ${field} : ""+${field}`};
+      return vgField(fieldOrDatumDef, {expr});
     }
+  } else {
+    return datumDefToExpr(fieldOrDatumDef);
   }
-  return undefined;
+}
+
+export function formatCustomType({
+  fieldOrDatumDef,
+  format,
+  formatType,
+  expr,
+  normalizeStack,
+  config,
+  field
+}: {
+  fieldOrDatumDef: FieldDef<string> | DatumDef<string>;
+  format: string | object;
+  formatType: string;
+  expr?: 'datum' | 'parent' | 'datum.datum';
+  normalizeStack?: boolean;
+  config: Config;
+  field?: string; // axis/legend "use datum.value"
+}) {
+  field = field ?? fieldToFormat(fieldOrDatumDef, expr, normalizeStack);
+
+  if (isFieldDef(fieldOrDatumDef) && isBinning(fieldOrDatumDef.bin)) {
+    const endField = vgField(fieldOrDatumDef, {expr, binSuffix: 'end'});
+    return {
+      signal: binFormatExpression(field, endField, format, formatType, config)
+    };
+  }
+  return {signal: customFormatExpr(formatType, field, format)};
+}
+
+export function formatGuide(
+  fieldOrDatumDef: FieldDef<string> | DatumDef<string>,
+  type: Type,
+  format: string | object,
+  formatType: string,
+  config: Config,
+  omitTimeFormatConfig: boolean // axis doesn't use config.timeFormat
+) {
+  if (isFieldOrDatumDefWithCustomTimeFormat(fieldOrDatumDef, config)) {
```

Comment:
What if the spec has a  custom `formatType` but config doesn't allow it?

It seems like we're not handling this here?  

(If you agree this is wrong, we can do this in a separate PR since I think previously we didn't handle this 100% correct either.)

Reply:
Let's handle this in a separate PR for https://github.com/vega/vega-lite/issues/6382